### PR TITLE
En 4289 limit p2p connections

### PR DESCRIPTION
**Pull request by traian-schiau**

### CODE REVIEW

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 1]
```diff
@@ -40,16 +44,62 @@ func (lcm *libp2pConnectionMonitor) Listen(network.Network, multiaddr.Multiaddr)
...
+	if lcm.targetConnCount > 0 {
+		return lcm.targetConnCount * 4 / 5
+	}
+	return 0
+}
+
+// ThresholdRandomTrim if the number of connected peers is over this value, we start cutting of connections at random
+func (lcm *libp2pConnectionMonitor) ThresholdRandomTrim() int {
+	if lcm.targetConnCount > 0 {
+		return lcm.targetConnCount * 6 / 5
```
Comment (bogdan-rosianu) [inline p2p/libp2p/libp2pConnectionMonitor.go]: extract this in p2p.toml as a percentage (`1.2` - float)

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 2]
```diff
@@ -40,16 +44,62 @@ func (lcm *libp2pConnectionMonitor) Listen(network.Network, multiaddr.Multiaddr)
...
+	if lcm.targetConnCount > 0 {
+		return lcm.targetConnCount
+	}
+	return math.MaxInt32
+}
+
+// ThresholdDiscoveryResume if the number of connected peers drop under this value, the dht discovery is restarted
+func (lcm *libp2pConnectionMonitor) ThresholdDiscoveryResume() int {
+	if lcm.targetConnCount > 0 {
+		return lcm.targetConnCount * 4 / 5
```
Comment (bogdan-rosianu) [inline p2p/libp2p/libp2pConnectionMonitor.go]: extract this in p2p.toml as a percentage (`0.8` - float)

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 3]
```diff
@@ -98,12 +101,37 @@ func (kdd *KadDhtDiscoverer) connectToInitialAndBootstrap() {
...
+
+					}
+				} else {
+					i++
+					if (i % 20) == 0 {
+						// KDD: Reconnect to initial list
+						chanRecInit := kdd.connectToOnePeerFromInitialPeersList(
+							kdd.refreshInterval,
+							kdd.initialPeersList)
+						<-chanRecInit
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Might reset variable i here as to always work with small values

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 4]
```diff
@@ -98,12 +101,37 @@ func (kdd *KadDhtDiscoverer) connectToInitialAndBootstrap() {
...
+						// KDD: no more peers, Reconnect to initial list
+						chanRecInit := kdd.connectToOnePeerFromInitialPeersList(
+							kdd.refreshInterval,
+							kdd.initialPeersList)
+						<-chanRecInit
+
+					}
+				} else {
+					i++
+					if (i % 20) == 0 {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Maybe extract 20 in a const

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 5]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
 	return kdd.connectToOnePeerFromInitialPeersList(kdd.refreshInterval, kdd.initialPeersList)
 }
 
+func (kdd *KadDhtDiscoverer) Pause() {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+	if kdd.initc {
+		// KDD: Pause kad-dht discovery
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: useless comment

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 6]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
 	return kdd.connectToOnePeerFromInitialPeersList(kdd.refreshInterval, kdd.initialPeersList)
 }
 
+func (kdd *KadDhtDiscoverer) Pause() {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Exported functions and variables should have comments

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 7]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
...
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+	if kdd.initc {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Why not set directly `kdd.initc = false`

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 8]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
...
+		// KDD: Pause kad-dht discovery
+		kdd.initc = false
+	}
+}
+
+func (kdd *KadDhtDiscoverer) Resume() {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+	if !kdd.initc {
+		// KDD: Resume kad-dht discovery
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: useless comment

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 9]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
...
+	if kdd.initc {
+		// KDD: Pause kad-dht discovery
+		kdd.initc = false
+	}
+}
+
+func (kdd *KadDhtDiscoverer) Resume() {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+	if !kdd.initc {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Why not set directly `kdd.initc = true`

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 10]
```diff
@@ -176,6 +204,24 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
...
+func (kdd *KadDhtDiscoverer) Pause() {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
+	if kdd.initc {
+		// KDD: Pause kad-dht discovery
+		kdd.initc = false
+	}
+}
+
+func (kdd *KadDhtDiscoverer) Resume() {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Exported functions and variables should have comments

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 11]
```diff
@@ -17,14 +19,16 @@ var ThresholdMinimumConnectedPeers = 3
...
 }
 
-func newLibp2pConnectionMonitor(reconnecter p2p.Reconnecter) *libp2pConnectionMonitor {
+func newLibp2pConnectionMonitor(reconnecter p2p.Reconnecter, targetConnCount int) *libp2pConnectionMonitor {
 	cm := &libp2pConnectionMonitor{
-		reconnecter:   reconnecter,
-		chDoReconnect: make(chan struct{}, 0),
+		reconnecter:     reconnecter,
+		chDoReconnect:   make(chan struct{}, 0),
+		targetConnCount: targetConnCount,
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply p2p/libp2p/libp2pConnectionMonitor.go]: In this case the checks in all other funcs will become useless
```GO
if lcm.targetConnCount > 0 {
     return lcm.targetConnCount
}
```

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 12]
```diff
@@ -40,16 +44,62 @@ func (lcm *libp2pConnectionMonitor) Listen(network.Network, multiaddr.Multiaddr)
...
+
+// ThresholdRandomTrim if the number of connected peers is over this value, we start cutting of connections at random
+func (lcm *libp2pConnectionMonitor) ThresholdRandomTrim() int {
+	if lcm.targetConnCount > 0 {
+		return lcm.targetConnCount * 6 / 5
+	}
+	return math.MaxInt32
+}
+
+// Request a reconnet to initial list
```
Comment (iulianpascalau) [inline p2p/libp2p/libp2pConnectionMonitor.go]: typo reconne(c)t

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 13]
```diff
@@ -40,16 +44,62 @@ func (lcm *libp2pConnectionMonitor) Listen(network.Network, multiaddr.Multiaddr)
...
 // Connected is called when a connection opened
-func (lcm *libp2pConnectionMonitor) Connected(network.Network, network.Conn) {}
+func (lcm *libp2pConnectionMonitor) Connected(netw network.Network, conn network.Conn) {
+	if len(netw.Conns()) > lcm.ThresholdDiscoveryPause() {
+		lcm.reconnecter.Pause()
+	}
+	if len(netw.Conns()) > lcm.ThresholdRandomTrim() {
+		sorted := kbucket.SortClosestPeers(netw.Peers(), kbucket.ConvertPeerID(netw.LocalPeer()))
+		for i := lcm.ThresholdDiscoveryPause(); i < len(sorted); i++ {
+			netw.ClosePeer(sorted[i])
```
Comment (iulianpascalau) [inline p2p/libp2p/libp2pConnectionMonitor.go]: returning error not handled

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 14]
```diff
@@ -40,16 +44,62 @@ func (lcm *libp2pConnectionMonitor) Listen(network.Network, multiaddr.Multiaddr)
...
+func (lcm *libp2pConnectionMonitor) doReconn() {
+	select {
+	case lcm.chDoReconnect <- struct{}{}:
+	default:
+	}
+}
+
 // Connected is called when a connection opened
-func (lcm *libp2pConnectionMonitor) Connected(network.Network, network.Conn) {}
+func (lcm *libp2pConnectionMonitor) Connected(netw network.Network, conn network.Conn) {
```
Comment (iulianpascalau) [inline p2p/libp2p/libp2pConnectionMonitor.go]: Should improve code coverage of the following funcs: Connected, ThresholdRandomTrim, ThresholdDiscoveryResume, ThresholdDiscoveryPause, Disconnected

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 15]
```diff
@@ -98,12 +101,37 @@ func (kdd *KadDhtDiscoverer) connectToInitialAndBootstrap() {
...
+					err := kdd.kadDHT.BootstrapOnce(ctx, cfg)
+					if err == kbucket.ErrLookupFailure {
+						// KDD: no more peers, Reconnect to initial list
+						chanRecInit := kdd.connectToOnePeerFromInitialPeersList(
+							kdd.refreshInterval,
+							kdd.initialPeersList)
+						<-chanRecInit
+
+					}
+				} else {
```
Comment (iulianpascalau) [inline p2p/libp2p/discovery/kadDhtDiscoverer.go]: Should improve code coverage on the else branch and in the functions: Pause, Resume

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 16]
```diff
@@ -27,6 +29,7 @@ type KadDhtDiscoverer struct {
 	refreshInterval  time.Duration
 	randezVous       string
 	initialPeersList []string
+	initc            bool
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply p2p/libp2p/discovery/kadDhtDiscoverer.go]: done

File: p2p/libp2p/discovery/kadDhtDiscoverer.go [hunk 17]
```diff
@@ -204,22 +201,25 @@ func (kdd *KadDhtDiscoverer) ReconnectToNetwork() <-chan struct{} {
...
 	kdd.mutKadDht.Lock()
 	defer kdd.mutKadDht.Unlock()
-	if !kdd.initc {
-		// KDD: Resume kad-dht discovery
-		kdd.initc = true
-	}
+	kdd.initConns = true
+}
+
+// IsDiscoveryPaused will return tru if the discoverer is initiating connections
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply p2p/libp2p/discovery/kadDhtDiscoverer.go]: done

File: p2p/libp2p/libp2pConnectionMonitor_test.go [hunk 18]
```diff
@@ -60,3 +62,84 @@ func TestNewLibp2pConnectionMonitor_OnDisconnectedUnderThresholdShouldCallReconn
...
+				*currentCount--
+				return nil
+			},
+		}
+	}
+
+	assert.Equal(t, 0, pauseCallCount)
+	assert.Equal(t, 0, resumeCallCount)
+
+	cm.Connected(netFact(5), nil)
```
Comment (iulianpascalau) [inline p2p/libp2p/libp2pConnectionMonitor_test.go]: We could have used a map for these kind of tests. Something like map[int][]int there the map key is the parameter sent to netFact function. The slice would have contained the expected values passed to assert.Equal function.

**Pull request merged**